src/audio/copier.c

```diff
@@ -0,0 +1,666 @@
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/ipc/msg.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/drivers/timer.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/cache.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/notifier.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc4/alh.h>
+#include <ipc4/copier.h>
+#include <ipc4/module.h>
+#include <ipc4/error_status.h>
+#include <ipc4/gateway.h>
+#include <ipc/dai.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_copier;
+
+/* this id aligns windows driver requirement to support windows driver */
+/* 9ba00c83-ca12-4a83-943c-1fa2e82f9dda */
+DECLARE_SOF_RT_UUID("copier", copier_comp_uuid, 0x9ba00c83, 0xca12, 0x4a83,
+		    0x94, 0x3c, 0x1f, 0xa2, 0xe8, 0x2f, 0x9d, 0xda);
+
+DECLARE_TR_CTX(copier_comp_tr, SOF_UUID(copier_comp_uuid), LOG_LEVEL_INFO);
+
+struct copier_data {
+	struct ipc4_copier_module_cfg config;
+	struct comp_dev *host;
+	struct comp_dev *dai;
+	int direction;
+
+	struct ipc4_audio_format out_fmt[IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT];
+	pcm_converter_func converter[IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT];
+};
+
+static enum sof_ipc_frame convert_fmt(int format)
+{
+	enum sof_ipc_frame in;
+
+	switch (format) {
+	case IPC4_DEPTH_16BIT:
+		in = SOF_IPC_FRAME_S16_LE;
+		break;
+	case IPC4_DEPTH_24BIT:
+		in = SOF_IPC_FRAME_S24_4LE;
+		break;
+	case IPC4_DEPTH_32BIT:
+		in = SOF_IPC_FRAME_S32_LE;
+		break;
+	default:
+		comp_cl_err(&comp_copier, "unsupported format %d", format);
+		return SOF_IPC_FRAME_S16_LE;
+	};
+
+	return in;
+}
+
+static struct comp_dev *create_host(struct comp_ipc_config *config,
+				    struct ipc4_copier_module_cfg *copier_cfg,
+				    int dir, struct pipeline *pipeline)
+{
+	struct sof_uuid host = {0x8b9d100c, 0x6d78, 0x418f, {0x90, 0xa3, 0xe0,
+			0xe8, 0x05, 0xd0, 0x85, 0x2b}};
+	struct ipc_config_host ipc_host;
+	const struct comp_driver *drv;
+	struct comp_dev *dev;
+
+	drv = ipc4_get_drv((uint8_t *)&host);
+	if (!drv)
+		return NULL;
+
+	config->type = SOF_COMP_HOST;
+
+	memset(&ipc_host, 0, sizeof(ipc_host));
+	ipc_host.direction = dir;
+
+	dev = drv->ops.create(drv, config, &ipc_host);
+	if (!dev)
+		return NULL;
+
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	return dev;
+}
+
+static void connect_comp_to_buffer(struct comp_dev *dev, struct comp_buffer *buf, int dir)
+{
+	uint32_t flags;
+
+	irq_local_disable(flags);
+	list_item_prepend(buffer_comp_list(buf, dir), comp_buffer_list(dev, dir));
+	buffer_set_comp(buf, dev, dir);
+	comp_writeback(dev);
+	irq_local_enable(flags);
+
+	dcache_writeback_invalidate_region(buf, sizeof(*buf));
+}
+
+static struct comp_dev *create_dai(struct comp_ipc_config *config,
```

Comment:
a comment here describing how the dai mode works.

Reply:
sure